java/semantickernel-api/src/main/java/com/microsoft/semantickernel/tokenizers/GPT3Tokenizer.java

```diff
@@ -0,0 +1,290 @@
+                    bpeTokens.add(GPT3Settings.encoder.get(encoding));
+                }
+            }
+        }
+
+        return bpeTokens;
+    }
+
+    public static List<Integer> encode(StringBuilder stringBuilder) {
+        if (stringBuilder != null) {
+            return encode(stringBuilder.toString());
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    public static List<Integer> encode(char[] chars) {
+        if (chars != null) {
+            return encode(new String(chars));
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    public static List<Integer> encode(Iterable<Character> chars) {
+        if (chars != null) {
+            StringBuilder sb = new StringBuilder();
+            for (Character c : chars) {
+                sb.append(c);
+            }
+            return encode(sb.toString());
+        } else {
+            return new ArrayList<>();
+        }
+    }
+
+    private static int encodingUtf8GetByteCount(String str) {
+        return str.getBytes(StandardCharsets.UTF_8).length;
+    }
+
+    private static int encodingUtf8GetBytes(String str, byte[] bytes) {
+        byte[] strBytes = str.getBytes(StandardCharsets.UTF_8);
+        System.arraycopy(strBytes, 0, bytes, 0, strBytes.length);
+        return strBytes.length;
+    }
+
+    private static List<String> bytePairEncoding(String token) {
+        if (bpeCache.containsKey(token)) {
+            return bpeCache.get(token);
+        }
+
+        if (token.length() <= 1) {
+            List<String> list = new ArrayList<>(1);
+            list.add(token);
+            bpeCache.put(token, list);
+            return list;
+        }
+
+        List<String> word = new ArrayList<>(token.length());
+        for (char c : token.toCharArray()) {
+            word.add(String.valueOf(c));
+        }
+
+        long smallestRank = Long.MAX_VALUE;
+        Tuple2<String, String> smallestPair = Tuples.of("", "");
+        List<String> newWord;
+
+        while (word.size() >= 2) {
+            for (int pairIndex = 0; pairIndex < word.size() - 1; pairIndex++) {
+                Tuple2<String, String> pair =
+                        Tuples.of(word.get(pairIndex), word.get(pairIndex + 1));
+
+                long pairRank =
+                        GPT3Settings.bpeRanks.containsKey(pair)
+                                ? GPT3Settings.bpeRanks.get(pair)
+                                : 100_000_000_000L;
+
+                if (pairRank <= smallestRank) {
+                    smallestRank = pairRank;
+                    smallestPair = pair;
+                }
+            }
+
+            if (!GPT3Settings.bpeRanks.containsKey(smallestPair)) {
+                break;
+            }
+
+            String first = smallestPair.getT1();
+            String second = smallestPair.getT2();
+
+            newWord = new ArrayList<>(word.size());
+            for (int i = 0; i < word.size(); i++) {
```

Comment:
I know this is a port, but I'd consider making this for loop a subroutine just to improve readability and shorten the method up. Not a requirement, and it is probably better to stick with the reference impl, so ignore this. :^)